File: app/src/main/java/com/wireguard/config/Interface.java
/*
 * Copyright © 2018 Samuel Holland <samuel@sholland.org>
 * Copyright © 2018 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.wireguard.config;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;
import androidx.annotation.Nullable;
import androidx.databinding.BaseObservable;
import androidx.databinding.Bindable;
import com.wireguard.android.Application;
import com.wireguard.android.BR;
import com.wireguard.android.R;
import com.wireguard.android.model.GlobalExclusions;
import com.wireguard.crypto.Keypair;

import java.net.InetAddress;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Represents the configuration for a WireGuard interface (an [Interface] block).
 */

public class Interface {
    private final List<InetNetwork> addressList;
    private final List<InetAddress> dnsList;
    private final List<String> excludedApplications;
    private final Context context = Application.Companion.get();
    @Nullable
    private Keypair keypair;
    private int listenPort;
    private int mtu;

    public Interface() {
        addressList = new ArrayList<>();
        dnsList = new ArrayList<>();
        excludedApplications = new ArrayList<>(Arrays.asList(Attribute.stringToList(GlobalExclusions.Companion.getExclusions())));
    }

    private void addAddresses(@Nullable final String[] addresses) {
        if (addresses != null && addresses.length > 0) {
            for (final String addr : addresses) {
                if (addr.isEmpty())
                    throw new IllegalArgumentException(context.getString(R.string.tunnel_error_empty_interface_address));
                addressList.add(new InetNetwork(addr));
            }
        }
    }

    private void addDnses(@Nullable final String[] dnses) {
        if (dnses != null && dnses.length > 0) {
            for (final String dns : dnses) {
                dnsList.add(InetAddresses.parse(dns));
            }
        }
    }

    private void addExcludedApplications(@Nullable final String[] applications) {
        if (applications != null && applications.length > 0) {
            excludedApplications.addAll(Arrays.asList(applications));
        }
    }

    @Nullable
    private String getAddressString() {
        if (addressList.isEmpty())
            return null;
        return Attribute.iterableToString(addressList);
    }

    private void setAddressString(@Nullable final String addressString) {
        addressList.clear();
        addAddresses(Attribute.stringToList(addressString));
    }

    public InetNetwork[] getAddresses() {
        return addressList.toArray(new InetNetwork[addressList.size()]);
    }

    @Nullable
    private String getDnsString() {
        if (dnsList.isEmpty())
            return null;
        return Attribute.iterableToString(getDnsStrings());
    }

    private void setDnsString(@Nullable final String dnsString) {
        dnsList.clear();
        addDnses(Attribute.stringToList(dnsString));
    }

    private List<String> getDnsStrings() {
        final List<String> strings = new ArrayList<>();
        for (final InetAddress addr : dnsList)
            strings.add(addr.getHostAddress());
        return strings;
    }

    public InetAddress[] getDnses() {
        return dnsList.toArray(new InetAddress[dnsList.size()]);
    }

    @Nullable
    private String getExcludedApplicationsString() {
        if (excludedApplications.isEmpty())
            return null;
        return Attribute.iterableToString(excludedApplications);
    }

    private void setExcludedApplicationsString(@Nullable final String applicationsString) {
        excludedApplications.clear();
        addExcludedApplications(Attribute.stringToList(applicationsString));
    }

    public String[] getExcludedApplications() {
        return excludedApplications.toArray(new String[excludedApplications.size()]);
    }

    public int getListenPort() {
        return listenPort;
    }

    private void setListenPort(final int listenPort) {
        this.listenPort = listenPort;
    }

    @Nullable
    private String getListenPortString() {
        if (listenPort == 0)
            return null;
        return Integer.valueOf(listenPort).toString();
    }

    private void setListenPortString(@Nullable final String port) {
        if (port != null && !port.isEmpty())
            setListenPort(Integer.parseInt(port, 10));
        else
            setListenPort(0);
    }

    public int getMtu() {
        return mtu;
    }

    private void setMtu(final int mtu) {
        this.mtu = mtu;
    }

    @Nullable
    private String getMtuString() {
        if (mtu == 0)
            return null;
        return Integer.toString(mtu);
    }

    private void setMtuString(@Nullable final String mtu) {
        if (mtu != null && !mtu.isEmpty())
            setMtu(Integer.parseInt(mtu, 10));
        else
            setMtu(0);
    }

    @Nullable
    public String getPrivateKey() {
        if (keypair == null)
            return null;
        return keypair.getPrivateKey();
    }

    private void setPrivateKey(@Nullable String privateKey) {
        if (privateKey != null && privateKey.isEmpty())
            privateKey = null;
        keypair = privateKey == null ? null : new Keypair(privateKey);
    }

    @Nullable
    public String getPublicKey() {
        if (keypair == null)
            return null;
        return keypair.getPublicKey();
    }

    public void parse(final String line) {
        final Attribute key = Attribute.match(line);
        if (key == null)
            throw new IllegalArgumentException(String.format(context.getString(R.string.tunnel_error_interface_parse_failed), line));
        switch (key) {
            case ADDRESS:
                addAddresses(key.parseList(line));
                break;
            case DNS:
                addDnses(key.parseList(line));
                break;
            case EXCLUDED_APPLICATIONS:
                addExcludedApplications(key.parseList(line));
                break;
            case LISTEN_PORT:
                setListenPortString(key.parse(line));
                break;
            case MTU:
                setMtuString(key.parse(line));
                break;
            case PRIVATE_KEY:
                setPrivateKey(key.parse(line));
                break;
            default:
                throw new IllegalArgumentException(line);
        }
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder().append("[Interface]\n");
        if (!addressList.isEmpty())
            sb.append(Attribute.ADDRESS.composeWith(addressList));
        if (!dnsList.isEmpty())
            sb.append(Attribute.DNS.composeWith(getDnsStrings()));
        if (!excludedApplications.isEmpty())
            sb.append(Attribute.EXCLUDED_APPLICATIONS.composeWith(excludedApplications));
        if (listenPort != 0)
            sb.append(Attribute.LISTEN_PORT.composeWith(listenPort));
        if (mtu != 0)
            sb.append(Attribute.MTU.composeWith(mtu));
        if (keypair != null)
            sb.append(Attribute.PRIVATE_KEY.composeWith(keypair.getPrivateKey()));
        return sb.toString();
    }

    public static class Observable extends BaseObservable implements Parcelable {
        public static final Creator<Observable> CREATOR = new Creator<Observable>() {
            @Override
            public Observable createFromParcel(final Parcel in) {
                return new Observable(in);
            }

            @Override
            public Observable[] newArray(final int size) {
                return new Observable[size];
            }
        };
        @Nullable
        private String addresses;
        @Nullable
        private String dnses;
        @Nullable
        private String excludedApplications;
        @Nullable
        private String listenPort;
        @Nullable
        private String mtu;
        @Nullable
        private String privateKey;
        @Nullable
        private String publicKey;

        public Observable(@Nullable final Interface parent) {
            if (parent != null)
                loadData(parent);
        }

        private Observable(final Parcel in) {
            addresses = in.readString();
            dnses = in.readString();
            publicKey = in.readString();
            privateKey = in.readString();
            listenPort = in.readString();
            mtu = in.readString();
            excludedApplications = in.readString();
        }

        public void commitData(final Interface parent) {
            parent.setAddressString(addresses);
            parent.setDnsString(dnses);
            parent.setExcludedApplicationsString(excludedApplications);
            parent.setPrivateKey(privateKey);
            parent.setListenPortString(listenPort);
            parent.setMtuString(mtu);
            loadData(parent);
            notifyChange();
        }

        @Override
        public int describeContents() {
            return 0;
        }

        public void generateKeypair() {
            final Keypair keypair = new Keypair();
            privateKey = keypair.getPrivateKey();
            publicKey = keypair.getPublicKey();
            notifyPropertyChanged(BR.privateKey);
            notifyPropertyChanged(BR.publicKey);
        }

        @Nullable
        @Bindable
        public String getAddresses() {
            return addresses;
        }

        public void setAddresses(final String addresses) {
            this.addresses = addresses;
            notifyPropertyChanged(BR.addresses);
        }

        @Nullable
        @Bindable
        public String getDnses() {
            return dnses;
        }

        public void setDnses(final String dnses) {
            this.dnses = dnses;
            notifyPropertyChanged(BR.dnses);
        }

        @Nullable
        @Bindable
        public String getExcludedApplications() {
            return excludedApplications;
        }

        public void setExcludedApplications(final String excludedApplications) {
            this.excludedApplications = excludedApplications;
            notifyPropertyChanged(BR.excludedApplications);
            notifyPropertyChanged(BR.excludedApplicationsCount);
        }

        @Bindable
        public int getExcludedApplicationsCount() {
            return Attribute.stringToList(excludedApplications).length;
        }

        @Nullable
        @Bindable
        public String getListenPort() {
            return listenPort;
        }

        public void setListenPort(final String listenPort) {
            this.listenPort = listenPort;
            notifyPropertyChanged(BR.listenPort);
        }

        @Nullable
        @Bindable
        public String getMtu() {
            return mtu;
        }

        public void setMtu(final String mtu) {
            this.mtu = mtu;
            notifyPropertyChanged(BR.mtu);
        }

        @Nullable
        @Bindable
        public String getPrivateKey() {
            return privateKey;
        }

        public void setPrivateKey(final String privateKey) {
            this.privateKey = privateKey;

            try {
                publicKey = new Keypair(privateKey).getPublicKey();
            } catch (final IllegalArgumentException ignored) {
                publicKey = "";
            }

            notifyPropertyChanged(BR.privateKey);
            notifyPropertyChanged(BR.publicKey);
        }

        @Nullable
        @Bindable
        public String getPublicKey() {
            return publicKey;
        }

        private void loadData(final Interface parent) {
            addresses = parent.getAddressString();
            dnses = parent.getDnsString();
            excludedApplications = parent.getExcludedApplicationsString();
            publicKey = parent.getPublicKey();
            privateKey = parent.getPrivateKey();
            listenPort = parent.getListenPortString();
            mtu = parent.getMtuString();
        }

        @Override
        public void writeToParcel(final Parcel dest, final int flags) {
            dest.writeString(addresses);
            dest.writeString(dnses);
            dest.writeString(publicKey);
            dest.writeString(privateKey);
            dest.writeString(listenPort);
            dest.writeString(mtu);
            dest.writeString(excludedApplications);
        }
    }
}

File: app/src/main/java/com/wireguard/config/Interface.kt
/*
 * Copyright © 2017-2018 WireGuard LLC. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.wireguard.config

import com.wireguard.android.util.ApplicationPreferences
import com.wireguard.android.util.requireNonNull
import com.wireguard.config.BadConfigException.Location
import com.wireguard.config.BadConfigException.Reason
import com.wireguard.config.BadConfigException.Section
import com.wireguard.crypto.Key
import com.wireguard.crypto.KeyFormatException
import com.wireguard.crypto.KeyPair
import java9.util.stream.Collectors
import java9.util.stream.StreamSupport
import java.net.InetAddress
import java.util.Collections
import java.util.LinkedHashSet
import java.util.Locale

/**
 * Represents the configuration for a WireGuard interface (an [Interface] block). Interfaces must
 * have a private key (used to initialize a `KeyPair`), and may optionally have several other
 * attributes.
 *
 *
 * Instances of this class are immutable.
 */
class Interface private constructor(builder: Builder) {

    /**
     * Returns the set of IP addresses assigned to the interface.
     *
     * @return a set of [InetNetwork]s
     */
    // The collection is already immutable.
    val addresses: Set<InetNetwork>
    /**
     * Returns the set of DNS servers associated with the interface.
     *
     * @return a set of [InetAddress]es
     */
    // The collection is already immutable.
    val dnsServers: Set<InetAddress>
    /**
     * Returns the set of applications excluded from using the interface.
     *
     * @return a set of package names
     */
    // This should ideally be immutable but I require it to not be
    // for my global exclusions implementation. Suggestions welcome
    // on alternate ways to do this.
    var excludedApplications: ArrayList<String>
    /**
     * Returns the public/private key pair used by the interface.
     *
     * @return a key pair
     */
    val keyPair: KeyPair
    /**
     * Returns the UDP port number that the WireGuard interface will listen on.
     *
     * @return a UDP port number, or `Optional.empty()` if none is configured
     */
    val listenPort: Int?
    /**
     * Returns the MTU used for the WireGuard interface.
     *
     * @return the MTU, or `Optional.empty()` if none is configured
     */
    val mtu: Int?

    init {
        // Defensively copy to ensure immutability even if the Builder is reused.
        addresses = Collections.unmodifiableSet(LinkedHashSet(builder.addresses))
        dnsServers = Collections.unmodifiableSet(LinkedHashSet(builder.dnsServers))
        excludedApplications = ArrayList(builder.excludedApplications)
        keyPair = builder.keyPair.requireNonNull("Interfaces must have a private key")
        listenPort = builder.listenPort
        mtu = builder.mtu
    }

    override fun equals(other: Any?): Boolean {
        if (other !is Interface)
            return false
        val obj = other as Interface?
        return (addresses == obj!!.addresses &&
            dnsServers == obj.dnsServers &&
            excludedApplications == obj.excludedApplications &&
            keyPair == obj.keyPair &&
            listenPort == obj.listenPort &&
            mtu == obj.mtu)
    }

    override fun hashCode(): Int {
        var hash = 1
        hash = 31 * hash + addresses.hashCode()
        hash = 31 * hash + dnsServers.hashCode()
        hash = 31 * hash + excludedApplications.hashCode()
        hash = 31 * hash + keyPair.hashCode()
        hash = 31 * hash + listenPort.hashCode()
        hash = 31 * hash + mtu.hashCode()
        return hash
    }

    /**
     * Converts the `Interface` into a string suitable for debugging purposes. The `Interface` is identified by its public key and (if set) the port used for its UDP socket.
     *
     * @return A concise single-line identifier for the `Interface`
     */
    override fun toString(): String {
        val sb = StringBuilder("(Interface ")
        sb.append(keyPair.publicKey.toBase64())
        listenPort?.let { lp -> sb.append(" @").append(lp) }
        sb.append(')')
        return sb.toString()
    }

    /**
     * Converts the `Interface` into a string suitable for inclusion in a `wg-quick`
     * configuration file.
     *
     * @return The `Interface` represented as a series of "Key = Value" lines
     */
    fun toWgQuickString(): String {
        val sb = StringBuilder()
        if (!addresses.isEmpty())
            sb.append("Address = ").append(Attribute.join(addresses)).append('\n')
        if (!dnsServers.isEmpty()) {
            val dnsServerStrings = StreamSupport.stream(dnsServers)
                .map<String>(InetAddress::getHostAddress)
                .collect(Collectors.toUnmodifiableList())
            sb.append("DNS = ").append(Attribute.join(dnsServerStrings)).append('\n')
        }
        if (!excludedApplications.isEmpty())
            sb.append("ExcludedApplications = ").append(
                Attribute.join(
                    excludedApplications + ApplicationPreferences.exclusionsArray
                )
            ).append('\n')
        listenPort?.let { lp -> sb.append("ListenPort = ").append(lp).append('\n') }
        mtu?.let { m -> sb.append("MTU = ").append(m).append('\n') }
        sb.append("PrivateKey = ").append(keyPair.privateKey.toBase64()).append('\n')
        return sb.toString()
    }

    /**
     * Serializes the `Interface` for use with the WireGuard cross-platform userspace API.
     * Note that not all attributes are included in this representation.
     *
     * @return the `Interface` represented as a series of "KEY=VALUE" lines
     */
    fun toWgUserspaceString(): String {
        val sb = StringBuilder()
        sb.append("private_key=").append(keyPair.privateKey.toHex()).append('\n')
        listenPort?.let { lp -> sb.append("listen_port=").append(lp).append('\n') }
        return sb.toString()
    }

    class Builder {
        // Defaults to an empty set.
        val addresses = LinkedHashSet<InetNetwork>()
        // Defaults to an empty set.
        val dnsServers = LinkedHashSet<InetAddress>()
        // Defaults to an empty set.
        val excludedApplications = LinkedHashSet<String>()
        // No default; must be provided before building.
        var keyPair: KeyPair? = null
        // Defaults to not present.
        var listenPort: Int? = null
        // Defaults to not present.
        var mtu: Int? = null

        fun addAddress(address: InetNetwork): Builder {
            addresses.add(address)
            return this
        }

        fun addAddresses(addresses: Collection<InetNetwork>): Builder {
            this.addresses.addAll(addresses)
            return this
        }

        fun addDnsServer(dnsServer: InetAddress): Builder {
            dnsServers.add(dnsServer)
            return this
        }

        fun addDnsServers(dnsServers: Collection<InetAddress>): Builder {
            this.dnsServers.addAll(dnsServers)
            return this
        }

        @Throws(BadConfigException::class)
        fun build(): Interface {
            if (keyPair == null)
                throw BadConfigException(
                    Section.INTERFACE, Location.PRIVATE_KEY,
                    Reason.MISSING_ATTRIBUTE, null
                )
            return Interface(this)
        }

        fun excludeApplication(application: String): Builder {
            excludedApplications.add(application)
            return this
        }

        fun excludeApplications(applications: Collection<String>): Builder {
            excludedApplications.addAll(applications)
            ApplicationPreferences.exclusionsArray.forEach { exclusion ->
                if (exclusion !in excludedApplications)
                    excludedApplications.add(exclusion)
            }
            return this
        }

        @Throws(BadConfigException::class)
        fun parseAddresses(addresses: CharSequence): Builder {
            try {
                for (address in Attribute.split(addresses))
                    addAddress(InetNetwork.parse(address))
                return this
            } catch (e: ParseException) {
                throw BadConfigException(Section.INTERFACE, Location.ADDRESS, e)
            }
        }

        @Throws(BadConfigException::class)
        fun parseDnsServers(dnsServers: CharSequence): Builder {
            try {
                for (dnsServer in Attribute.split(dnsServers))
                    addDnsServer(InetAddresses.parse(dnsServer))
                return this
            } catch (e: ParseException) {
                throw BadConfigException(Section.INTERFACE, Location.DNS, e)
            }
        }

        fun parseExcludedApplications(apps: CharSequence): Builder {
            return excludeApplications(
                Attribute.split(apps)
                    .filter { it !in ApplicationPreferences.exclusionsArray }
                    .toList()
            )
        }

        @Throws(BadConfigException::class)
        fun parseListenPort(listenPort: String): Builder {
            try {
                return setListenPort(Integer.parseInt(listenPort))
            } catch (e: NumberFormatException) {
                throw BadConfigException(Section.INTERFACE, Location.LISTEN_PORT, listenPort, e)
            }
        }

        @Throws(BadConfigException::class)
        fun parseMtu(mtu: String): Builder {
            try {
                return setMtu(Integer.parseInt(mtu))
            } catch (e: NumberFormatException) {
                throw BadConfigException(Section.INTERFACE, Location.MTU, mtu, e)
            }
        }

        @Throws(BadConfigException::class)
        fun parsePrivateKey(privateKey: String): Builder {
            try {
                return setKeyPair(KeyPair(Key.fromBase64(privateKey)))
            } catch (e: KeyFormatException) {
                throw BadConfigException(Section.INTERFACE, Location.PRIVATE_KEY, e)
            }
        }

        fun setKeyPair(keyPair: KeyPair): Builder {
            this.keyPair = keyPair
            return this
        }

        @Throws(BadConfigException::class)
        fun setListenPort(listenPort: Int): Builder {
            if (listenPort < MIN_UDP_PORT || listenPort > MAX_UDP_PORT)
                throw BadConfigException(
                    Section.INTERFACE, Location.LISTEN_PORT,
                    Reason.INVALID_VALUE, listenPort.toString()
                )
            this.listenPort = if (listenPort == 0) 0 else listenPort
            return this
        }

        @Throws(BadConfigException::class)
        fun setMtu(mtu: Int): Builder {
            if (mtu < 0)
                throw BadConfigException(
                    Section.INTERFACE, Location.LISTEN_PORT,
                    Reason.INVALID_VALUE, mtu.toString()
                )
            this.mtu = if (mtu == 0) 0 else mtu
            return this
        }
    }

    companion object {
        private const val MAX_UDP_PORT = 65535
        private const val MIN_UDP_PORT = 0

        /**
         * Parses an series of "KEY = VALUE" lines into an `Interface`. Throws
         * [ParseException] if the input is not well-formed or contains unknown attributes.
         *
         * @param lines An iterable sequence of lines, containing at least a private key attribute
         * @return An `Interface` with all of the attributes from `lines` set
         */
        @Throws(BadConfigException::class)
        fun parse(lines: Iterable<CharSequence>): Interface {
            val builder = Builder()
            for (line in lines) {
                val attribute = Attribute.parse(line) ?: throw
                    BadConfigException(
                        Section.INTERFACE, Location.TOP_LEVEL,
                        Reason.SYNTAX_ERROR, line
                    )
                when (attribute.key.toLowerCase(Locale.ENGLISH)) {
                    "address" -> builder.parseAddresses(attribute.value)
                    "dns" -> builder.parseDnsServers(attribute.value)
                    "excludedapplications" -> builder.parseExcludedApplications(attribute.value)
                    "listenport" -> builder.parseListenPort(attribute.value)
                    "mtu" -> builder.parseMtu(attribute.value)
                    "privatekey" -> builder.parsePrivateKey(attribute.value)
                    else -> throw BadConfigException(
                        Section.INTERFACE, Location.TOP_LEVEL,
                        Reason.UNKNOWN_ATTRIBUTE, attribute.key
                    )
                }
            }
            return builder.build()
        }
    }
}